src/poprox_recommender/evaluation/evaluate.py

```diff
@@ -0,0 +1,116 @@
+import csv
+import gzip
+import json
+import logging
+
+import numpy as np
+import pandas as pd
+from lenskit.metrics import topn
+from tqdm import tqdm
+
+from poprox_concepts.api.recommendations import RecommendationRequest
+from poprox_recommender.data.mind import MindData
+from poprox_recommender.evaluation.metrics import rank_biased_overlap
+from poprox_recommender.paths import project_root
+
+logger = logging.getLogger("poprox_recommender.measurement_offline")
+
+
+def compute_rec_metric(recs_df: pd.DataFrame, request: RecommendationRequest):
+    # Make sure you run generate.py first to get the recommendations data file
+    recs = recs_df[recs_df["user"] == request.interest_profile.profile_id]
```

Comment:
This is the first thing to speed up. For every test user (360K+), it is doing a linear scan over the entire recs data frame, so this process is O(N^2).

If you move ``user`` to the index in the main code, before your loop:

```python
recs_df = recs_df.set_index('user', append=True)
```

Then you can do a fast lookup:

```python
recs = recs_df.loc[str(request.interest_profile.profile_id)]
```

Reply:
Yeah after the change now the time is lowered to 12 hr! I'll keep exploring if there can be other optimization

Reply:
Yep, there definitely will be.